runtime/interpreter/interpreter.go

```diff
@@ -2214,8 +2217,182 @@ func (interpreter *Interpreter) writeStored(
 	interpreter.recordStorageMutation()
 }
 
+type FromStringFunctionValue struct {
+	parentType   sema.Type
+	hostFunction *HostFunctionValue
+}
+
+func fromStringDeclaration(ty sema.Type, impl func(*Interpreter, string) Value) FromStringFunctionValue {
+	functionType, _ := sema.FromStringMetadata(ty)
+
+	hostFunctionImpl := NewUnmeteredHostFunctionValue(
+		func(invocation Invocation) Value {
+			argument, ok := invocation.Arguments[0].(*StringValue)
+			if !ok {
+				// expect typechecker to catch a mismatch here
+				panic(errors.NewUnreachableError())
+			}
+			inter := invocation.Interpreter
+			return impl(inter, argument.Str)
+		},
+		functionType,
+	)
+	return FromStringFunctionValue{
+		parentType:   ty,
+		hostFunction: hostFunctionImpl,
+	}
+}
+
+// default implementation for parsing a given unsigned numeric type from a string.
+// the size provided by sizeInBytes is passed to strconv.ParseUint, ensuring that the parsed value fits in the target type.
+// input strings must not begin with a '+' or '-'.
+func fromStringImplUnsigned(bitSize int, convert func(uint64) Value) func(*Interpreter, string) Value {
+	expectedUsage := common.NewNumberMemoryUsage(bitSize / 8)
+
+	return func(interpreter *Interpreter, input string) Value {
+		val, err := strconv.ParseUint(input, 10, bitSize)
+		if err != nil {
+			return NewNilValue(interpreter)
+		}
+
+		converted := convert(val)
+		common.UseMemory(interpreter, expectedUsage)
+		return NewSomeValueNonCopying(interpreter, converted)
+	}
+}
+
+// default implementation for parsing a given signed numeric type from a string.
+// the size provided by sizeInBytes is passed to strconv.ParseUint, ensuring that the parsed value fits in the target type.
+// input strings may begin with a '+' or '-'.
+func fromStringImplSigned(bitSize int, convert func(int64) Value) func(*Interpreter, string) Value {
+	expectedUsage := common.NewCadenceBigIntMemoryUsage(bitSize / 8)
+
+	return func(interpreter *Interpreter, input string) Value {
+		val, err := strconv.ParseInt(input, 10, bitSize)
+		if err != nil {
+			return NewNilValue(interpreter)
+		}
+
+		converted := convert(val)
+		common.UseMemory(interpreter, expectedUsage)
+		return NewSomeValueNonCopying(interpreter, converted)
+	}
+}
+
+func fromStringImplBig(convert func(*big.Int) (Value, bool)) func(*Interpreter, string) Value {
+	return func(interpreter *Interpreter, input string) Value {
+		val, ok := common.NewBigInt(interpreter).SetString(input, 10)
```

Comment:
Creating the big int allocates, so needs to be metered before

Reply:
Could you elaborate a bit more? It seems that `common.NewBigInt` meters an uninitialized bigInt, and `SetString` returns nil on failure. Should we meter the state of `val` after the parsing step, regardless of success?

The docs for `setString` also state that "the value of z is undefined" if parsing fails.

Reply:
Metering, like memory metering or computation metering, must always occur *before* the operation, so that the metering to prevent the future operation if it is determined that the limit would be exceeded.

Basically, move the `common.UseMemory` up before this line and over-estimate the amount of memory based on the length of the string.